In [1]:
import pandas as pd
import numpy as np

In [2]:
filename = './athleteEventsNoPersonal.csv'
oly_df = pd.read_csv(filename)
oly_df.info()
oly_df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40616 entries, 0 to 40615
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      40616 non-null  int64  
 1   Age     39143 non-null  float64
 2   Height  31615 non-null  float64
 3   Team    40616 non-null  object 
 4   NOC     40616 non-null  object 
 5   Games   40616 non-null  object 
 6   Year    40616 non-null  int64  
 7   Season  40616 non-null  object 
 8   City    40616 non-null  object 
 9   Sport   40616 non-null  object 
 10  Event   40616 non-null  object 
 11  Medal   5917 non-null   object 
dtypes: float64(2), int64(2), object(8)
memory usage: 3.7+ MB


ID   Age  Height         Team  NOC        Games  Year  Season         City  \
0   5  21.0   185.0  Netherlands  NED  1988 Winter  1988  Winter      Calgary   
1   5  21.0   185.0  Netherlands  NED  1988 Winter  1988  Winter      Calgary   
2   5  25.0   185.0  Netherlands  NED  1992 Winter  1992  Winter  Albertville   
3   5  25.0   185.0  Netherlands  NED  1992 Winter  1992  Winter  Albertville   
4   5  27.0   185.0  Netherlands  NED  1994 Winter  1994  Winter  Lillehammer   

           Sport                               Event Medal  
0  Speed Skating    Speed Skating Women's 500 metres   NaN  
1  Speed Skating  Speed Skating Women's 1,000 metres   NaN  
2  Speed Skating    Speed Skating Women's 500 metres   NaN  
3  Speed Skating  Speed Skating Women's 1,000 metres   NaN  
4  Speed Skating    Speed Skating Women's 500 metres   NaN

In [3]:
# displays rows and columns
oly_df.shape
print(f"There are {oly_df.shape[0]} rows and {oly_df.shape[1]} columns in this dataset")

There are 40616 rows and 12 columns in this dataset


In [4]:
# 207 duplicates
oly_df.duplicated().sum()

207

In [5]:
oly_df = oly_df.drop_duplicates()
oly_df.duplicated().sum()

0

In [6]:
oly_df.isna().sum()

ID            0
Age        1446
Height     8796
Team          0
NOC           0
Games         0
Year          0
Season        0
City          0
Sport         0
Event         0
Medal     34493
dtype: int64

In [7]:
# practicing filter
london_city = oly_df['City'] == 'London'
oly_df[london_city]

ID   Age  Height     Team  NOC        Games  Year  Season    City  \
36         54   NaN     NaN     Iran  IRI  1948 Summer  1948  Summer  London   
42         67  26.0   179.0   Russia  RUS  2012 Summer  2012  Summer  London   
66        155  34.0   185.0    Egypt  EGY  2012 Summer  2012  Summer  London   
69        159  19.0   167.0    Egypt  EGY  2012 Summer  2012  Summer  London   
72        185  31.0     NaN    Egypt  EGY  1948 Summer  1948  Summer  London   
...       ...   ...     ...      ...  ...          ...   ...     ...     ...   
40577  135489  22.0   182.0   Russia  RUS  2012 Summer  2012  Summer  London   
40578  135489  22.0   182.0   Russia  RUS  2012 Summer  2012  Summer  London   
40590  135517  22.0     NaN  Austria  AUT  1948 Summer  1948  Summer  London   
40595  135528  28.0   181.0  Germany  GER  2012 Summer  2012  Summer  London   
40609  135558  32.0   200.0   Poland  POL  2012 Summer  2012  Summer  London   

                       Sport                                         Event  \
36                    Boxing                      Boxing Men's Heavyweight   
42                 Athletics               Athletics Women's Javelin Throw   
66                 Wrestling      Wrestling Men's Heavyweight, Greco-Roman   
69     Synchronized Swimming            Synchronized Swimming Women's Team   
72                   Fencing                Fencing Men's Foil, Individual   
...                      ...                                           ...   
40577               Swimming        Swimming Women's 200 metres Backstroke   
40578               Swimming  Swimming Women's 4 x 100 metres Medley Relay   
40590               Swimming          Swimming Men's 100 metres Backstroke   
40595              Badminton                       Badminton Men's Singles   
40609             Volleyball                   Volleyball Men's Volleyball   

        Medal  
36        NaN  
42        NaN  
66        NaN  
69        NaN  
72        NaN  
...       ...  
40577  Silver  
40578     NaN  
40590     NaN  
40595     NaN  
40609     NaN  

[3347 rows x 12 columns]

     a. How many rows? 40616

     b. How many columns? 12

     c. Which columns have missing values? (Why do think there are there so many null values in the Medals column?) The most missing values are in the "Medal" column because not everyone who enters in the Olympics wins a metal. 
 
     d. How many entries correspond to the city of London? 3347

     e. What age is the youngest athlete in our sample data?  Hint: use min() 11

In [20]:
# oly_df['Sport'].unique()
tug_war = oly_df['Sport'] == 'Tug-Of-War'
# oly_df[tug_war]
team_swed = oly_df['Team'] == "Sweden"
# oly_df[team_swed]

# swed_tug =  oly_df.loc[(oly_df['Sport'] == 'Tug-Of-War') & (oly_df['Team'] == "Sweden")]
swed_tug = oly_df.loc[(tug_war) & (team_swed)]
swed_tug


ID   Age  Height    Team  NOC        Games  Year  Season    City  \
10023  34205  22.0     NaN  Sweden  SWE  1908 Summer  1908  Summer  London   
12681  43105  23.0     NaN  Sweden  SWE  1906 Summer  1906  Summer  Athina   
13015  44087  29.0   182.0  Sweden  SWE  1906 Summer  1906  Summer  Athina   
16250  55169  23.0     NaN  Sweden  SWE  1908 Summer  1908  Summer  London   
19048  64266  21.0     NaN  Sweden  SWE  1908 Summer  1908  Summer  London   
20550  68647  26.0   190.0  Sweden  SWE  1906 Summer  1906  Summer  Athina   

            Sport                        Event   Medal  
10023  Tug-Of-War  Tug-Of-War Men's Tug-Of-War     NaN  
12681  Tug-Of-War  Tug-Of-War Men's Tug-Of-War  Bronze  
13015  Tug-Of-War  Tug-Of-War Men's Tug-Of-War  Bronze  
16250  Tug-Of-War  Tug-Of-War Men's Tug-Of-War     NaN  
19048  Tug-Of-War  Tug-Of-War Men's Tug-Of-War     NaN  
20550  Tug-Of-War  Tug-Of-War Men's Tug-Of-War  Bronze

In [25]:
gold_metals = oly_df['Medal'] == 'Gold'
# oly_df[gold_metals]
over_40 = oly_df['Age'] >= 40
# oly_df[over_40]
gold_40 = oly_df[(gold_metals) & (over_40)]
gold_40

ID   Age  Height                Team  NOC        Games  Year  \
740      2735  41.0   168.0              Russia  RUS  2000 Summer  2000   
1103     4029  46.0     NaN           Argentina  ARG  1936 Summer  1936   
1421     4993  41.0   180.0             Mosk II  NOR  1920 Summer  1920   
3040    10769  49.0   180.0          White Lady  DEN  1964 Summer  1964   
3270    11617  44.0     NaN             Heroine  GBR  1908 Summer  1908   
...       ...   ...     ...                 ...  ...          ...   ...   
38916  130271  43.0     NaN            Heira II  NOR  1920 Summer  1920   
38952  130381  42.0   167.0       United States  USA  1972 Summer  1972   
39193  130936  56.0     NaN       United States  USA  1908 Summer  1908   
39197  130936  60.0     NaN       United States  USA  1912 Summer  1912   
39365  131533  58.0     NaN  Cincinnati Archers  USA  1904 Summer  1904   

       Season       City             Sport  \
740    Summer     Sydney          Shooting   
1103   Summer     Berlin              Polo   
1421   Summer  Antwerpen           Sailing   
3040   Summer      Tokyo           Sailing   
3270   Summer     London           Sailing   
...       ...        ...               ...   
38916  Summer  Antwerpen           Sailing   
38952  Summer     Munich           Archery   
39193  Summer     London          Shooting   
39197  Summer  Stockholm  Art Competitions   
39365  Summer  St. Louis           Archery   

                                             Event Medal  
740    Shooting Men's Rapid-Fire Pistol, 25 metres  Gold  
1103                               Polo Men's Polo  Gold  
1421                       Sailing Mixed 10 metres  Gold  
3040           Sailing Mixed Three Person Keelboat  Gold  
3270                        Sailing Mixed 7 metres  Gold  
...                                            ...   ...  
38916                      Sailing Mixed 12 metres  Gold  
38952                   Archery Women's Individual  Gold  
39193   Shooting Men's Running Target, Double Shot  Gold  
39197           Art Competitions Mixed Sculpturing  Gold  
39365                   Archery Women's Team Round  Gold  

[76 rows x 12 columns]

2. Create filters to find out:

     a. How many athletes who participated in the Sport "Tug-Of-War" were from Team "Sweden" in our sample dataset? 6 athletes 

     b. How many Gold medals were awarded to athletes who were over 40 years old? 76 gold metals
     

In [28]:
# oly_df['Height'].describe()
oly_df['Height'].mean()


175.3812988327587

In [33]:
oly_df.groupby(['Event'])['Height'].mean() # grouped by 1 column and seeing another column's mean
# oly_df.groupby(['Event','Height']).mean() can group by multiple columns

Event
Alpine Skiing Men's Combined                      178.547170
Alpine Skiing Men's Downhill                      178.211679
Alpine Skiing Men's Giant Slalom                  176.739394
Alpine Skiing Men's Slalom                        176.482143
Alpine Skiing Men's Super G                       179.850575
                                                     ...    
Wrestling Women's Flyweight, Freestyle            158.100000
Wrestling Women's Heavyweight, Freestyle          173.444444
Wrestling Women's Light-Heavyweight, Freestyle    170.000000
Wrestling Women's Lightweight, Freestyle          162.666667
Wrestling Women's Middleweight, Freestyle         164.533333
Name: Height, Length: 726, dtype: float64

In [34]:
oly_single_athlete_df = oly_df.copy() #copy df

In [42]:
oly_single_athlete_df = oly_single_athlete_df.drop_duplicates(subset = 'ID', keep='first')
# oly_single_athlete_df['ID'].unique
oly_single_athlete_df['Height'].mean()

176.31818479042698

3. Determine the average height if:

     a. All rows are included.

     b. All rows are included but grouped by Event.

     c. Bonus: This dataset contains many athletes more than once (if they competed in multiple events and/or multiple years). You can confirm this by finding  .value_counts on the ID column!  For the bonus task we want to get the average height of all  the athletes, but don't want to include any athlete more than once.  Hint: We learned how to drop duplicates if EVERY column was identical, but now we want to drop a row any time just the ID is repeated.  To do this,  you can use a subset of your data as an argument in the drop_duplicates function.  You can also tell Python which of the duplicates to keep by using the keep argument.